policy-controller/grpc/src/lib.rs

```diff
@@ -182,9 +182,53 @@ where
             })?;
 
         self.index
-            .service_lookup(addr, port)
+            .lookup_ip(addr, port)
             .ok_or_else(|| tonic::Status::not_found("No such service"))
     }
+
+    fn lookup_authority(
+        &self,
+        authority: &str,
+    ) -> Result<(String, String, NonZeroU16), tonic::Status> {
+        let auth = authority
+            .parse::<http::uri::Authority>()
+            .map_err(|_| tonic::Status::invalid_argument("invalid authority"))?;
+
+        let mut host = auth.host();
+        if host.is_empty() {
+            return Err(tonic::Status::invalid_argument(
+                "authority must have a host",
+            ));
+        }
+        if host.ends_with('.') {
+            host = &host[..host.len() - 1];
+        }
+        if host.ends_with(&*self.cluster_domain) {
+            host = &host[..host.len() - self.cluster_domain.len()];
+        }
+        let mut parts = host.split('.');
+        let invalid = {
+            let domain = &self.cluster_domain;
+            move || {
+                tonic::Status::invalid_argument(format!(
+                    "authority must be of the form <name>.<namespace>.svc.{}",
+                    domain
+                ))
+            }
+        };
+        let name = parts.next().ok_or_else(invalid)?;
+        let namespace = parts.next().ok_or_else(invalid)?;
+        if !matches!(parts.next(), Some("svc")) {
```

Comment:
nit: _pretty_ sure the `matches!` is unnecessary here since `Option<&str>` implements `Eq`?
```suggestion
        if parts.next() != Some("svc") {
```
or did that not work?